### ISSUE TITLE
[Bug]: Next.js and rules_ts are incompatible for custom tsconfig.json options

### ISSUE THREAD

**Issue by christian-roggia** [aspect-build/bazel-examples#226]
### What happened?

Next.js is currently trying to overwrite the `tsconfig.json` at runtime with its "suggested" values. Some options are unfortunately incompatible with `rules_ts`, specifically the configuration `noEmit` must be `false` or `undefined` for the rules to work but Next.js attempts to overwrite it with the value `true` enabling it.

In the `js_run_devserver` the attribute `grant_sandbox_write_permissions` can be enabled as a workaround to let Next.js do its magic. Unfortunately, this is not possible in `js_run_binary` rule.

The options are therefore the following:

- Embed in the srcs the `tsconfig.json` with `noEmit` disabled. This will fail as Next.js attempts to overwrite the read-only file.
- Embed in the srcs the `tsconfig.json` with `noEmit` enabled. This will fail as `rules_ts` require output files to be generated (`noEmit` must be disabled).
- Do not embed in the srcs the `tsconfig.json` and let Next.js generate a default one during runtime. This is a suboptimal option as no custom configuration can be defined (this is the current approach in the Next.js example in this repository).
- Embed in the srcs a runtime `tsconfig.json` with `noEmit` enabled, other compile options must be compatible with Next.js enforced options. Reference a second, different `tsconfig.build.json` during build time (i.e. in the `ts_project` definitions under the `tsconfig` attribute) which is as close as possible to the runtime configuration. This is suboptimal but the best solution found so far.

NOTE: Next.js does not allow disabling overwriting the `tsconfig.json`.
NOTE: Next.js does not allow to specify a different `tsconfig.json`, it must be in the root of the project being compiled.

Keywords:

- `Error: EROFS: read-only file system, open`
- `cannot be built because the 'noEmit' option is specified in the tsconfig`

### Version

-

**Comment by gregmagolan** [comment]
Unlike `js_run_devserver` which creates its own custom sandbox where it can control the permissions of the files, `js_run_binary` rule runs in the bazel output tree where Bazel controls the permissions of the files. It would be not idiomatic to change the permissions of the files there and it could cause issues with other actions that are running in parallel and also using the tsconfig.json file.

The least hacky fix here would be to patch Next.js in your repository to disable the tsconfig writes. I've seen this done at a client site.

**Comment by christian-roggia** [comment]
@gregmagolan I believe that Next.js should have addressed this issue a long time ago, I found dozens of closed PRs and issues mentioning the nightmare of dealing with Next.js actively overwriting values in that `tsconfig.json`. I find it pretty sad that all attempts to solve this have been discarded, in multiple iterations they considered only printing a warning instead of rewriting the file, that would be the optimal solution here.

For context I added a working example of how to use two almost identical `tsconfig.json` that allow for custom options with `noEmit` as the only difference. My advice for anyone adopting this workaround is to keep the two configurations as close as possible.

**Comment by alexeagle** [comment]
Yeah Bazel can only do what the underlying tool allows. I'd suggest filing more issues with Next.js so they know their approach causes incompatibilities. I think they can default to this sort of user-friendly behavior but should have a flag to turn it off so their tool can be composed into other build systems.

The next.js example moved to https://github.com/bazelbuild/examples/tree/main/frontend - please file an issue there if you think there's something we should change in the example code .- thanks
